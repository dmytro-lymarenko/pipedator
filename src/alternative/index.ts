import { createValidator, getFirstErrors, Validator, getCurrentPath } from '../core';

export function alternative<ValidValue = any>(validators: Validator[], message?: string) {
	return createValidator<ValidValue>({
		validate: (value, ctx) => {
			if (validators.length === 0) {
				return null;
			}
			// find the first success validator
			const errors = getFirstErrors(i => validators[i], () => value, () => ctx, validators.length);

			if (errors.length === validators.length) {
				// here all validators returned an error
				if (errors.length === 1) {
					return errors[0];
				}

				return {
					path: getCurrentPath(ctx),
					message: message || `Either ${errors.map(error => error.message).join(' OR ')}`,
					children: null,
				};
			}

			return null;
		},
	});
}
