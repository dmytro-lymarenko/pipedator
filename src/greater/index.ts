import { createValidator, ValidationRef, getCurrentPath } from '../core';

export function greater<ValidValue = any>(limit: number | ValidationRef, message?: string) {
	return createValidator<ValidValue>({
		validate: (value, ctx) => {
			const v = typeof limit === 'number' ? limit : limit(ctx);

			return value > v
				? null
				: {
						path: getCurrentPath(ctx),
						message: message || `Value should be greater than ${v}`,
						children: null,
				  };
		},
	});
}
