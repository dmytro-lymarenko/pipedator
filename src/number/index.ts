import { test } from '../test';

export function number<ValidValue = any>(message?: string) {
	return test<ValidValue>(value => typeof value === 'number', message || 'Value should be a number');
}
