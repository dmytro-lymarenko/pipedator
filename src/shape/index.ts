/**
 * @module validators
 */

import { Validator } from '../core';
import { abstractShape } from '../abstractShape';

export function shape(shape: { [key: string]: Validator }, options?: { onlyFirstError?: boolean }, message?: string) {
	const keys = Object.keys(shape);

	return abstractShape(keys, key => shape[key], options, message);
}
