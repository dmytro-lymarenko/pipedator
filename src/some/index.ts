import { createValidator, getFirstRequirements, Validator } from '../core';
import { alternativeRequirementFactory } from '../core/requirements';

/**
 * value should be an array
 * @param validator
 */
export function some<ValidValue = any>(validator: Validator, message?: string) {
	return createValidator<ValidValue>({
		validate: (value, ctx) => {
			if (!Array.isArray(value)) {
				throw new Error('Value should be an array');
			}

			if (value.length === 0) {
				return null;
			}
			// find the first success validator
			const requirements = getFirstRequirements(
				() => validator,
				i => value[i],
				i => ({
					...ctx,
					path: [...ctx.path, i.toString()],
				}),
				value.length
			);

			if (requirements.length === value.length) {
				// here all values failed
				const requirement = requirements[0];

				return alternativeRequirementFactory(
					message || `At least one item should follow the rule: ${requirement.message}`,
					requirements
				)(ctx.path, value);
			}

			return null;
		},
	});
}
