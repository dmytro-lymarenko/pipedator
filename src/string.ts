import { createValidator } from './core';

export function string() {
	return createValidator({
		message: 'Value should be a string',
		validate: (value, ctx) => (typeof value === 'string' ? null : { value, message: ctx.message, path: ctx.path }),
	});
}

export function empty() {
	return createValidator({
		message: 'Value should be empty',
		validate: (value, ctx) => (value.length === 0 ? null : { value, message: ctx.message, path: ctx.path }),
	});
}
