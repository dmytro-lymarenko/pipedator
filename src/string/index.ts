import { test } from '../test';

export function string<ValidValue = any>(message?: string) {
	return test<ValidValue>(value => typeof value === 'string', message || 'Value should be a string');
}
