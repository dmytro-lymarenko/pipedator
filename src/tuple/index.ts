import { Validator } from '../core';
import { abstractShape, AbstractShapeOptions } from '../abstractShape';

export function tuple(tuple: Validator[], options?: AbstractShapeOptions, message?: string) {
	const keys = tuple.map((_, i) => i);

	return abstractShape(keys, key => tuple[key], options, message);
}
