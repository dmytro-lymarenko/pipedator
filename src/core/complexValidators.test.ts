import { number } from '../number';
import { shape } from '../shape';
import { string } from '../string';
import { some } from '../some';
import { tuple } from '../tuple';
import { alternative } from '../alternative';
import { greater } from '../greater';

describe('complex validators', () => {
	it('should work', () => {
		expect(
			some(
				shape({
					foo: string(),
				})
			).validate([null, { foo: 2 }])
		).toMatchSnapshot();

		expect(
			some(
				shape({
					foo: shape({
						bar: string(),
					}),
					valid: number(),
				})
			).validate([{ foo: 1, valid: 1 }])
		).toMatchSnapshot();

		expect(
			some(
				tuple([
					shape({
						bar: string(),
					}),
					number(),
				])
			).validate([1, 1])
		).toMatchSnapshot();

		expect(alternative(string(), alternative(number(), greater(5))).validate({})).toMatchSnapshot();

		expect(alternative(string(), number()).validate({})).toMatchSnapshot();
	});
});
