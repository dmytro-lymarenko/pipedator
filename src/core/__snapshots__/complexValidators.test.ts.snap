// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex validators should work 1`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "message": "Value should be defined",
      "path": "0",
    },
    Object {
      "children": Array [
        Object {
          "children": null,
          "message": "Value should be a string",
          "path": "foo",
        },
      ],
      "message": "Value should have a valid shape",
      "path": "1",
    },
  ],
  "message": "At least one child should be valid",
  "path": "",
}
`;

exports[`complex validators should work 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": null,
              "message": "Value should be a string",
              "path": "bar",
            },
          ],
          "message": "Value should have a valid shape",
          "path": "foo",
        },
      ],
      "message": "Value should have a valid shape",
      "path": "0",
    },
  ],
  "message": "At least one child should be valid",
  "path": "",
}
`;

exports[`complex validators should work 3`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": null,
          "message": "Value should be defined",
          "path": "0",
        },
        Object {
          "children": null,
          "message": "Value should be a number",
          "path": "1",
        },
      ],
      "message": "Value should have a valid shape",
      "path": "0",
    },
    Object {
      "children": Array [
        Object {
          "children": null,
          "message": "Value should be defined",
          "path": "0",
        },
        Object {
          "children": null,
          "message": "Value should be a number",
          "path": "1",
        },
      ],
      "message": "Value should have a valid shape",
      "path": "1",
    },
  ],
  "message": "At least one child should be valid",
  "path": "",
}
`;

exports[`complex validators should work 4`] = `
Object {
  "children": null,
  "message": "Either Value should be a string OR Either Value should be a number OR Value should be greater than 5",
  "path": "",
}
`;

exports[`complex validators should work 5`] = `
Object {
  "children": null,
  "message": "Either Value should be a string OR Value should be a number",
  "path": "",
}
`;
